### PR TITLE
Add transformer for web app manifests

### LINKED ISSUE #235: 🐛 Problem with loading manifest.json
### 🐛 Bug report
While building or running in dev mode with parcel, if I include manifest.json in html, then parcel return's me .js file path.

### 🤔 Expected Behavior

```html
<link rel="manifest" href="/dist/manifest.json">
```

### 😯 Current Behavior

```html
<link rel="manifest" href="/dist/05bfa35095e6bfacdb8b334d1023872b.js">
```

### 🌍 Your Environment

| Software         | Version(s)
| ---------------- | ----------
| Parcel           |  1.2.0
| Node             | node 8.9.2
| npm/Yarn         | yarn 1.3.2
| Operating System |  macOS Sierra 10.12.6

### LINKED ISSUE #301: 🙋 PWA support
_This is both a question and a :raising_hand_woman: feature request._

__Question:__ <br>
Since a PWA requires a set of resources to exists along side its `manifest.webmanifest` file. My question is, is there a way to include a file without having its name "scrambled"? This would apply to both the manifest file and its resources. _(Related to https://github.com/parcel-bundler/parcel/issues/280)_

__Feature request:__ <br>
It would be nice to have parcel look through the manifest file for files to include in the build.
For example (partial files):
```html
<!-- index.html -->
<link rel="manifest" href="./manifest.webmanifest">
``` 
```js
// manifest.webmanifest
"icons": [{
      "src": "pwa/icons/48.png",
      "sizes": "48x48",
      "type": "image/png"
    }, {
      "src": "pwa/icons/72.png",
      "sizes": "72x72",
      "type": "image/png"
    }
]
```
Should result in both `manifest.webmanifest` as well as `pwa/icons/48.png` and `pwa/icons/72.png` being included in the build.

### LINKED ISSUE #3362: Parcel 2: Web Manifest transformer
Create the `@parcel/transformer-webmanifest` package in `packages/transforms/webmanifest`. This package is a Parcel 2 transform plugin for web manifests. It should be based on the `WebManifestAsset` from Parcel 1.

The `@parcel/transformer-webmanifest` package is responsible for the following things:

* Parsing JSON
* Collecting dependencies
* Rewriting urls
* Generating JSON

We currently only handle the `.webmanifest` extension, but many people want to be able to use `.json` as well (#235, #1039). In order to support this, we need a way for the HTML transformer to specify that the dependency should be treated as a `webmanifest` file regardless of its actual extension.

### PR DESCRIPTION
**Pull request by ranisalt**
# ↪️ Pull Request

Considering that #4313 is stale, has merge conflicts and a blocking issue, I rewrote it to avoid address those problems.

In contrast with the mentioned issue, this uses the `.webmanifest` extension instead of `webmanifest:` key just because it is simpler.

Closes #4313, closes #3404, fixes #3362, fixes #301

## 💻 Examples

See `packages/core/integration-tests/test/integration/webmanifest/` for a working example, I decided to give it a try since I really need this feature. I've been developing and using [a standalone version of this](https://github.com/ranisalt/parcel-transformer-webmanifest/) but to have it built-in would be great.

## ✔️ PR Todo

- [x] Added/updated ~unit~ integration tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

File: packages/transformers/webmanifest/src/WebManifestTransformer.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
+// @flow
+// https://developer.mozilla.org/en-US/docs/Web/Manifest
+import type {MutableAsset} from '@parcel/types';
+import type {SchemaEntity} from '@parcel/utils';
+
+import invariant from 'assert';
+import {parse} from 'json-source-map';
```
Comment (mischnic) [inline packages/transformers/webmanifest/src/WebManifestTransformer.js:7]: You need to do a default import and access `.parse` on there.

File: packages/transformers/webmanifest/src/WebManifestTransformer.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    const source = await asset.getCode();
+    const {data, pointers} = parse(source);
+
+    validateSchema.diagnostic(
+      MANIFEST_SCHEMA,
+      {source, map: {data, pointers}, filePath: asset.filePath},
+      // asset.filePath,
+      // source,
+      '@parcel/transformer-webmanifest',
+      // `/${encodeJSONKeyComponent('@parcel/transformer-webmanifest')}`,
```
Comment (mischnic) [inline packages/transformers/webmanifest/src/WebManifestTransformer.js:48]: Can these comments be removed?

**Comment by ranisalt** [comment]
Suggestions applied

**Comment by mischnic** [comment]
This needs to be changed to get Flow to pass:
```diff
diff --git flow-libs/json-source-map.js.flow flow-libs/json-source-map.js.flow
index 62133567..eda1c8a1 100644
--- flow-libs/json-source-map.js.flow
+++ flow-libs/json-source-map.js.flow
@@ -25,7 +25,7 @@ declare module 'json-source-map' {
       _?: mixed,
       opts?: {|bigint: boolean|},
     ) => {|
-      data: mixed,
+      data: any,
       pointers: {|[key: string]: Mapping|},
     |},
```

File: packages/core/integration-tests/test/webmanifest.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            message,
+            filePath: manifestPath,
+            origin: '@parcel/core',
+          },
+          {
+            hints: [],
+            message: `Cannot load file './icon.png' in '${path.dirname(
+              manifestPathRelative,
+            )}'.`,
+            // origin: '@parcel/resolver-default',
```
Comment (mischnic) [inline packages/core/integration-tests/test/webmanifest.js:148]: The fix for that is
```patch
diff --git packages/core/core/src/requests/PathRequest.js packages/core/core/src/requests/PathRequest.js
index eacced75..3492e4b6 100644
--- packages/core/core/src/requests/PathRequest.js
+++ packages/core/core/src/requests/PathRequest.js
@@ -211,11 +211,11 @@ export class ResolverRunner {
           }

           if (result.diagnostics) {
-            if (Array.isArray(result.diagnostics)) {
-              diagnostics.push(...result.diagnostics);
-            } else {
-              diagnostics.push(result.diagnostics);
-            }
+            let errorDiagnostic = errorToDiagnostic(
+              new ThrowableDiagnostic({diagnostic: result.diagnostics}),
+              resolver.name,
+            );
+            diagnostics.push(...errorDiagnostic);
           }
         }
       } catch (e) {
```

After that, you should be able to uncomment the line.

Comment (ranisalt) [reply packages/core/integration-tests/test/webmanifest.js]: Sure! I usually refrain from modifying other files in pull requests but I guess this is needed.

File: packages/core/core/src/requests/PathRequest.js [hunk 4]
```diff
@@ -1,7 +1,6 @@
 // @flow strict-local
 import type {Diagnostic} from '@parcel/diagnostic';
 import type {Async, QueryParameters} from '@parcel/types';
-import type {StaticRunOpts} from '../RequestTracker';
```
Comment (mischnic) [inline packages/core/core/src/requests/PathRequest.js:4]: That line shouldn't removed

**Review by mischnic — APPROVED** [review]

**Comment by mischnic** [comment]
> In contrast with the mentioned issue, this uses the .webmanifest extension instead of webmanifest: key just because it is simpler.

The one downside with this is that `manifest.json`doesn't work. A lot of people complained about that in https://github.com/parcel-bundler/parcel/issues/235. I don't really see a reason again `*.webmanifest` though...

**Comment by ranisalt** [comment]
@mischnic yes, that was my first idea, however, I decided it was just too hard to support `manifest.json` without it conflicting with the JSON utilities. `*.webmanifest` is accepted by the spec and there's no reason to specifically use `manifest.json` as the file name.

**Comment by mischnic** [comment]
Yes, but the whole issue https://github.com/parcel-bundler/parcel/issues/235 is about why `manifest.json` doesn't work.

Either this was documented badly or they really don't want to use `.webmanifest` 🤷

**Comment by ranisalt** [comment]
> Either this was documented badly or they really don't want to use `.webmanifest` shrug

I think we can have this functionality only for `*.webmanifest` files until someone figures out how to allow for `manifest.json` to be used too. Better have something than nothing, or even, better to have a "why can't I use an arbitraty filename?" report than a "why this feature is unavailable?" :stuck_out_tongue_closed_eyes:

**Comment by damianobarbati** [comment]
@ranisalt is this supposed to be working? I'm on `"parcel": "^2.0.0-beta.1"`.

My .parcelrc:
```json
{
    "extends": "@parcel/config-default",
    "transformers": {
        "*.png": ["@parcel/transformer-raw"]
    }
}
```

Screenshot 2021-05-02 at 18 12 55

**Comment by ranisalt** [comment]
@damianobarbati update your parcel to the latest beta

**Comment by damianobarbati** [comment]
@ranisalt I'm on (from yarn.lock)
```
=> Found "parcel@2.0.0-beta.2"
```
Is it correct?

**Comment by mischnic** [comment]
Yes, you seem to have disabled autoinstall.

So `yarn add -D @parcel/transformer-webmanifest@2.0.0-beta.2`

**Comment by damianobarbati** [comment]
That made it, thanks!

**Pull request merged**